text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+
+* `Peer`s: other nodes in the network (other Bees and Hornets),
+* `Message`es: the information exchanged between peers, and its serialization/deserialization.
+
+Given some identifier `epid`, sending a message to its corresponding endpoint becomes a single line of asynchronous code:
+
+```rust
+network.send(SendMessage { epid, bytes: "hello".as_bytes() }).await?;
+```
+
+The purpose of this crate is to provide the following functionalities:
+* maintain a list of endpoints (not peers!),
+* establish and maintain connections with endpoints,
+* allow to send, multicast or broadcast byte encoded messages of variable size to any of the endpoints,
+* allow to receive byte encoded messages of variable size from any of the endpoints,
+* reject connections from unknown, i.e. not whitelisted, endpoints,
+* manage all connections and message transfers asynchronously,
+* provide an extensible, yet convenient-to-use, and documented API.
+
+The key design decisions are now being discussed in the following sub-sections.
+
+## async/await
+
+This crate is by nature very much dependent on events happening outside of the control of the program, e.g. listening for incoming connections from peers, waiting for packets on a specific socket, etc. Hence, this crate under the hood makes heavy use of Rust's concurrency abstractions. Luckily, with the stabilization of the `async/await` syntax, writing asynchronous code has become almost as easy to read, write, and maintain as synchronous code. In an experimental implementation of this crate the [`async_std`](https://github.com/async-rs/async-std) library was used, which comes with asynchronous drop-in replacements for their synchronous equivalents found in Rust's standard library `std`. Additionally, asynchronous mpsc channels were taken from the [`futures`](https://github.com/rust-lang/futures-rs) crate.
+
+## Message passing
+
+This crate favors message passing via channels over globally shared state and locks. Instead of keeping the list of endpoints in a globally accessible hashmap this crate separates and moves such state into workers that run asynchronously, and are listening for commands and events to act upon, and also notify listeners by sending events.
+
+## Initialization
+
+In order for this crate to be used, it has to be initialized first by providing a `NetworkConfig`:
+
+```rust
+struct NetworkConfig {
+    // The server or binding TCP port that remote endpoints can connect and send to.
+    binding_port: u16,
+    // The binding address that remote endpoints can connect and send to.
+    binding_addr: IpAddr,
+    // The interval between two connections attempts in case of connection losses.
+    reconnect_interval: u64,
+    // The maximum message length in terms of bytes.
+    max_message_length: usize
```

Comment:
oh, I see now that the term might be confusing. It's just the buffer size where received bytes will be stored in. We this to be allocated only once. So in the current protocol this number has to be set to 1604, which is the worst case size a message that would fill the buffer completely.

Reply:
Alright but shouldn't it be a constant then ? If it's a conf we can only assume it *will* be modified. What happens if it's too small then ?

Reply:
I am not sure what happens internally if a message with more than 1604 bytes is sent. I would assume a resize (and therefore reallocation), but could also just discard the extant bytes.

Reply:
`Alright but shouldn't it be a constant then ?` I would not make it constant, because then people who'ld like to use it would have to edit and recompile that crate. I think it's fair to assume, that if someone wants to use this crate for his purposes he'ld like to be able to change the buffer size depending on his/her usecase.

Reply:
Fair enough. As soon as we're sure of the behaviours on an insufficient buffer I'm fine.